terra-notebooks-playground/Hail - How to run a notebook in the background.ipynb

```diff
@@ -0,0 +1,224 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Setup"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import nbformat\n",
+    "from nbconvert.preprocessors import CellExecutionError\n",
+    "from nbconvert.preprocessors import ExecutePreprocessor\n",
+    "\n",
+    "import os\n",
+    "import time"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "<div class=\"alert alert-block alert-warning\">\n",
+    "<b>Change the cell below</b> to hold the name of the notebook in this workspace that you wish to run in the background. There is no need to modify any of the other cells.\n",
+    "</div>"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "#---------[ CHANGE THIS TO BE THE NAME OF THE NOTEBOOK YOU WISH TO RUN ]-----------\n",
+    "NOTEBOOK_TO_RUN = 'your_notebook.ipynb'"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Programmatically set output paths"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Formulate a new filename for the output notebook so that it include a date and timestamp for when it was executed."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "TIMESTAMP_FILE_SUFFIX = time.strftime('_%Y%m%d_%H%M%S.ipynb')\n",
+    "OUTPUT_NOTEBOOK = NOTEBOOK_TO_RUN.replace('.ipynb', TIMESTAMP_FILE_SUFFIX)\n",
+    "\n",
+    "print(f'Executed notebook will be written to filename \"{OUTPUT_NOTEBOOK}\" on the local disk and the workspace bucket.')"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "DATESTAMP = time.strftime('%Y%m%d')\n",
+    "HAIL_LOG_DIR_FOR_PROVENANCE = f'{os.getenv(\"WORKSPACE_BUCKET\")}/hail-logs/{DATESTAMP}/'\n",
+    "\n",
+    "print(f'Hail logs will be copied to {HAIL_LOG_DIR_FOR_PROVENANCE}')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Execute the notebook and capture provenance"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "<div class=\"alert alert-block alert-info\">\n",
+    "The next cell will use <kbd>nbconvert</kbd> to run the notebook in the background until it completes (or yields an error) and will capture all notebook outputs as a separate notebook file.\n",
+    "</div>"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# See also https://nbconvert.readthedocs.io/en/latest/execute_api.html\n",
+    "with open(NOTEBOOK_TO_RUN) as f_in:\n",
+    "    nb = nbformat.read(f_in, as_version=4)\n",
```

Comment:
Curious, do you know what version 4 refers to? I see examples in documentation that refer to version 4, but I don't understand what it is.

Reply:
Its the version of the format of the ipynb file https://nbformat.readthedocs.io/en/latest/format_description.html.

Looks like the only major versions are 4 and 5 right now per https://nbformat.readthedocs.io/en/latest/changelog.html